tools/wpt/browser.py

```diff
@@ -516,264 +516,201 @@ def version(self, binary=None, webdriver_binary=None):
             raise ValueError("Unable to construct a valid Chrome package name for current platform")
-        platform = platform.lower()
 
-        if platform == "linux":
+        if self.platform == "Linux":
             bits = "64" if uname[4] == "x86_64" else "32"
-        elif platform == "mac":
+        elif self.platform == "Mac":
             bits = "64"
-        elif platform == "win":
+        elif self.platform == "Win":
             bits = "32"
 
-        return "%s%s" % (platform, bits)
+        return f"{self.platform.lower()}{bits}"
 
     def _chromium_platform_string(self):
-        platform = self.platforms.get(uname[0])
-
-        if platform is None:
-            raise ValueError("Unable to construct a valid Chromium package name for current platform")
+        if self.platform is None:
+            raise ValueError(
+                "Unable to construct a valid Chromium package name for current platform")
 
-        if (platform == "Linux" or platform == "Win") and uname[4] == "x86_64":
-            platform += "_x64"
+        if (self.platform == "Linux" or self.platform == "Win") and uname[4] == "x86_64":
+            return f"{self.platform}_x64"
 
-        return platform
+        return self.platform
 
     def _chromium_package_name(self):
-        return "chrome-%s" % self.platforms.get(uname[0]).lower()
+        return f"chrome-{self.platform.lower()}"
 
-    def _latest_chromium_snapshot_url(self):
-        # Make sure we use the same revision in an invocation.
+    def _format_chromium_snapshot_url(self, revision=None):
         architecture = self._chromium_platform_string()
-        if self._last_change is None:
-            revision_url = "https://storage.googleapis.com/chromium-browser-snapshots/%s/LAST_CHANGE" % architecture
-            self._last_change = get(revision_url).text.strip()
-        return "https://storage.googleapis.com/chromium-browser-snapshots/%s/%s/" % (architecture, self._last_change)
+        # Make sure we use the same revision in an invocation.
+        # If we have a last revision used, that revision takes priority.
+        if self.last_revision_used is None and revision is None:
+            # No revision provided or previously used. use latest revision.
+            self.revision_attempted = self._get_latest_chromium_revision(architecture)
+        elif revision is not None:
+            self.revision_attempted = revision
+        else:
+            self.revision_attempted = self.last_revision_used
 
-    def find_nightly_binary(self, dest):
-        if uname[0] == "Darwin":
-            return find_executable("Chromium",
-                                   os.path.join(dest, self._chromium_package_name(), "Chromium.app", "Contents", "MacOS"))
-        # find_executable will add .exe on Windows automatically.
-        return find_executable("chrome", os.path.join(dest, self._chromium_package_name()))
+        # The attempted revision is kept. If it is successfully used,
```

Comment:
It seems like it will be used again even if not successful, is that right? Setting it before it is attempted to be used makes this name a bit funny. Could we have a flow like this?
- Figure out what binaries we want to try to download
- Download those binaries
- Record in the filesystem what exact version those binaries are
- Cache those versions in the Python code if we need to look them up many times during an invocation of `./wpt run`

Reply:
I changed some wording to make the logic a bit more readable, but I'll also try to explain. The way it currently works, a revision number will be determined for use and kept in the `revision_attempted` attribute. 
https://github.com/web-platform-tests/wpt/blob/c841a6b1b9dbf77fc1aa5a4ad41a88f148751467/tools/wpt/browser.py#L567-L582

After all other steps are successful (formulating the url, trying to download, unzip, and install the file), then `last_revision_used` is changed to reference the attempted revision number, and that number always takes precedent in any other installs with no revision number passed (browser binary, webdriver, mojojs).
This happens after the browser binary successfully installs:
https://github.com/web-platform-tests/wpt/blob/c841a6b1b9dbf77fc1aa5a4ad41a88f148751467/tools/wpt/browser.py#L711-L716

Or the webdriver successfully installs:
https://github.com/web-platform-tests/wpt/blob/c841a6b1b9dbf77fc1aa5a4ad41a88f148751467/tools/wpt/browser.py#L814-L821

However, If the install attempt had failed, the attempted revision will not be kept in `last_revision_used` for reuse. This is the implementation I came up with to avoid trying to reuse a revision number that had previously failed. I do see that it might read as confusing. I originally only had a `last_revision_used` attribute that was set when the url is created, but it's not unusual to attempt to use a url to download, fail, and then generate another (usually with the latest revision instead).  I wanted to implement a way that the revision is only reused after everything was successful. This is what I came up with, but I'm up for changing it to a better approach.

Reply:
After explaining all this, I reimplemented this to not keep revisions attempted and successfully used as attributes, and instead just keep the url that was successfully used for reuse.
https://github.com/web-platform-tests/wpt/blob/615e384f4863ee297d4fc6afde94f02cb5c90b7b/tools/wpt/browser.py#L732-L736
Hopefully this is more intuitive.